README
Net-Xero

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it to get an idea of the module's uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Net::Xero

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Net-Xero

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Net-Xero

    CPAN Ratings
        http://cpanratings.perl.org/d/Net-Xero

    Search CPAN
        http://search.cpan.org/dist/Net-Xero/


COPYRIGHT AND LICENCE

Copyright (C) 2010 Lenz Gschwendtner

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Net-Xero

0.20    2012-05-31T01:37:23
        proper utf8 support
        added reference field to payments template

0.19    2012-04-23T14:12:55
        more bugfixing. uri_escape still not doing exactly what we need

0.18    2012-04-20T12:41:41
        Bugfix release: fixed encoding issue with uri_escape

0.17    2012-04-12T00:19:52
        added ManualJournal support

0.16    2012-03-07T22:44:55
        added support for search of multiple invoice refs with one call

0.15    2012-02-29T21:56:42
        clear old error message before making next call

0.14    2011-12-10T11:12:13
        removed calculated values from templates, let xero do this

0.13    2011-12-09T19:41:12
        major rewrite of internal _talk() method, simplified
        fixed reading cert from file

0.12    2011-12-02T22:25:34
        sign post data before sending

0.11    2011-12-02T12:27:54
        upated to conform with latest API changes and URLs

0.10    2011-01-13T11:53:01
        tested payments and working template

0.9     2010-12-07T12:58:16
        added payments command

0.8     2010-12-01T12:56:06
        change the template path to the dist_dir

0.7     2010-11-29T17:34:47
        per line item tax definition if present

0.6     2010-11-17T10:22:46
        fixed the template path

0.5     2010-11-16T18:27:44
        added set_cert method

0.4     2010-11-15T10:02:26
        fixed package building

0.3     2010-11-11T15:56:30
        changed the API to make it easier to extend

0.2     2010-11-11T14:36:48
        fixed Makefile to install the shared directory

0.1     2010-11-11T14:07:25
        first version of the lib that can add invoices and credit notes

0.01    Date/time
        First version, released on an unsuspecting world.



t/test.pl
#!/usr/bin/env perl
#

use common::sense;
use Net::Xero;
use Data::Dumper;

my $xro = Net::Xero->new();
$xro->debug(1);
$xro->set_cert('xero.key');
$xro->key('xxx');
$xro->secret('yyy');
$xro->access_token('xxx');
$xro->access_secret('yyy');

my $xero_id;
eval { $xero_id = $xro->put('items', { item_id => 'wtf123' }); };

if ($@ or $xro->error) {
    my $error = $@ || $xro->error;
    die "error while talking to xero: " . $error;
}

print Dumper($xero_id);


lib/Net/Xero.pm
package Net::Xero;

use 5.010;
use Mouse;
use Net::OAuth;
use LWP::UserAgent;
use HTTP::Request::Common;
use Data::Random qw(rand_chars);
use XML::LibXML::Simple qw(XMLin);
use File::ShareDir 'dist_dir';
use Template;
use Crypt::OpenSSL::RSA;
use URI::Escape;
use Data::Dumper;

$Net::OAuth::PROTOCOL_VERSION = Net::OAuth::PROTOCOL_VERSION_1_0A;

=head1 NAME

Net::Xero - The great new Net::Xero!

=head1 VERSION

Version 0.20

=cut

our $VERSION = '0.20';

has 'api_url' => (
    is      => 'rw',
    isa     => 'Str',
    default => 'https://api.xero.com',
);
has 'ua' => (
    is      => 'rw',
    isa     => 'LWP::UserAgent',
    default => sub { new LWP::UserAgent },
);
has 'debug' => (
    is      => 'rw',
    isa     => 'Bool',
    default => 0,
);
has 'error' => (
    is        => 'rw',
    isa       => 'Str',
    predicate => 'has_error',
    clearer   => 'clear_error',
);
has 'key'    => (is => 'rw', isa => 'Str');
has 'secret' => (is => 'rw', isa => 'Str');
has 'cert'   => (is => 'rw', isa => 'Str');
has 'nonce'  => (
    is      => 'ro',
    isa     => 'Str',
    default => join('', rand_chars(size => 16, set => 'alphanumeric')),
);
has 'login_link' => (is => 'rw', isa => 'Str');
has 'callback_url' => (
    is      => 'rw',
    isa     => 'Str',
    default => 'http://localhost:3000/callback',
);
has 'request_token'  => (is => 'rw', isa => 'Str');
has 'request_secret' => (is => 'rw', isa => 'Str');
has 'access_token'   => (is => 'rw', isa => 'Str');
has 'access_secret'  => (is => 'rw', isa => 'Str');
has 'template_path'  => (
    is      => 'rw',
    isa     => 'Str',
    default => (-d dist_dir('Net-Xero') ? dist_dir('Net-Xero') : {}),
);

#has 'template_path' => (is => 'rw', isa => 'Str');

=head1 SYNOPSIS

Quick summary of what the module does.

Perhaps a little code snippet.

    use Net::Xero;

    my $foo = Net::Xero->new();
    ...

=head1 EXPORT

A list of functions that can be exported.  You can delete this section
if you don't export anything, such as for a purely object-oriented module.

=head1 FUNCTIONS

=cut

=head2 login

This sets up the initial OAuth handshake and returns the login URL. This
URL has to be clicked by the user and the the user then has to accept
the application in xero.

Xero then redirects back to the callback URL defined with
C<$self-E<gt>callback_url>. If the user already accepted the application the
redirect may happen without the user actually clicking anywhere.

=cut

sub login {
    my $self = shift;

    my $request = Net::OAuth->request("request token")->new(
        consumer_key     => $self->key,
        consumer_secret  => $self->secret,
        request_url      => $self->api_url . '/oauth/RequestToken',
        request_method   => 'POST',
        signature_method => 'RSA-SHA1',
        timestamp        => time,
        nonce            => $self->nonce,
        callback         => $self->callback_url,
    );

    my $private_key = Crypt::OpenSSL::RSA->new_private_key($self->cert);
    $request->sign($private_key);
    my $res = $self->ua->request(GET $request->to_url);

    if ($res->is_success) {
        my $response =
            Net::OAuth->response('request token')
            ->from_post_body($res->content);
        $self->request_token($response->token);
        $self->request_secret($response->token_secret);
        print STDERR "Got Request Token ", $response->token, "\n"
            if $self->debug;
        print STDERR "Got Request Token Secret ", $response->token_secret, "\n"
            if $self->debug;
        return
              $self->api_url
            . '/oauth/Authorize?oauth_token='
            . $response->token
            . '&oauth_callback='
            . $self->callback_url;
    }
    else {
        $self->error($res->status_line);
        warn "Something went wrong: " . $res->status_line;
    }
}

=head2 auth

The auth method changes the initial request token into access token that we need
for subsequent access to the API. This method only has to be called once
after login.

=cut

sub auth {
    my $self = shift;

    my $request = Net::OAuth->request("access token")->new(
        consumer_key     => $self->key,
        consumer_secret  => $self->secret,
        request_url      => $self->api_url . '/oauth/AccessToken',
        request_method   => 'POST',
        signature_method => 'RSA-SHA1',
        timestamp        => time,
        nonce            => $self->nonce,
        callback         => $self->callback_url,
        token            => $self->request_token,
        token_secret     => $self->request_secret,
    );
    my $private_key = Crypt::OpenSSL::RSA->new_private_key($self->cert);
    $request->sign($private_key);
    my $res = $self->ua->request(GET $request->to_url);

    if ($res->is_success) {
        my $response =
            Net::OAuth->response('access token')->from_post_body($res->content);
        $self->access_token($response->token);
        $self->access_secret($response->token_secret);
        print STDERR "Got Access Token ", $response->token, "\n"
            if $self->debug;
        print STDERR "Got Access Token Secret ", $response->token_secret, "\n"
            if $self->debug;
    }
    else {
        $self->error($res->status_line);
        warn "Something went wrong: " . $res->status_line;
    }
}

=head2 set_cert

=cut

sub set_cert {
    my ($self, $path) = @_;

    unless (-f $path) {
        warn("No such file: $path: " . $!);
        return;
    }

    open(CERT, '<', $path)
        or warn("Could not read certificate: $path: " . $!);
    $self->cert(join('', <CERT>));
    close CERT;
}

=head2 get_inv_by_ref

=cut

sub get_inv_by_ref {
    my ($self, @ref) = @_;

    my $path = 'Invoices?where=Reference.ToString()=="' . (shift @ref) . '"';
    $path .= ' OR Reference.ToString()=="' . $_ . '"' foreach (@ref);

    return $self->_talk($path, 'GET');
}

=head2 get_invoices

=cut

sub get_invoices {
    my ($self, $where) = @_;

    my $path = 'Invoices';

    return $self->_talk($path, 'GET') unless (ref $where eq 'HASH');

    $path .= '?where=';
    my $conjunction =
        (exists $where->{'conjunction'}) ? uc $where->{'conjunction'} : 'OR';
    my $first = 1;

    foreach my $key (%{$where}) {
        $path .= " $conjunction " unless $first;

        given ($key) {
            when ('reference') {
                my @refs = @{ $where->{$key} };
                $path .= 'Reference.ToString()=="' . (shift @refs) . '"';
                $path .= ' OR Reference.ToString()=="' . $_ . '"'
                    foreach (@refs);
            }
            when ('contact') {
                my @contacts = @{ $where->{$key} };
                my $contact  = shift @contacts;
                $path .= join(
                    ' AND ',
                    map {
                              "Contact."
                            . ucfirst($_) . '=="'
                            . $contact->{$_} . '"'
                        } keys %{$contact});

                # finish foreach
            }
            when ('number') {
                my @numbers = @{ $where->{$key} };
                $path .= ' OR InvoiceNumber.ToString()=="' . $_ . '"'
                    foreach (@numbers);
            }
        }

        $first = 0;
    }

    return $self->_talk($path, 'GET');
}

=head2 create_invoice

=cut

sub create_invoice {
    my ($self, $hash) = @_;
    $hash->{command} = 'create_invoice';
    return $self->_talk('Invoices', 'POST', $hash);
}

=head2 approve_credit_note

=cut

sub approve_credit_note {
    my ($self, $hash) = @_;
    $hash->{command} = 'approve_credit_note';
    return $self->_talk('CreditNotes', 'POST', $hash);
}

=head2 status_invoice

=cut

sub status_invoice {
    my ($self, $hash) = @_;
    $hash->{command} = 'status_invoice';
    return $self->_talk('Invoices', 'POST', $hash);
}

=head2 get

=cut

sub get {
    my ($self, $command) = @_;
    return $self->_talk($command, 'GET');
}

=head2 post

=cut

sub post {
    my ($self, $command, $hash) = @_;
    return $self->_talk($command, 'POST', $hash);
}

=head2 put

=cut

sub put {
    my ($self, $command, $hash) = @_;
    return $self->_talk($command, 'PUT', $hash);
}

=head1 INTERNAL API

=head2 _talk

_talk handles the access to the restricted resources. You should
normally not need to access this directly.

=cut

sub _talk {
    my ($self, $command, $method, $hash) = @_;

    $self->clear_error;

    my $path = join('', map(ucfirst, split(/_/, $command)));

    my $request_url = $self->api_url . '/api.xro/2.0/' . $path;
    my %opts        = (
        consumer_key     => $self->key,
        consumer_secret  => $self->secret,
        request_url      => $request_url,
        request_method   => $method,
        signature_method => 'RSA-SHA1',
        timestamp        => time,
        nonce        => join('', rand_chars(size => 16, set => 'alphanumeric')),
        token        => $self->access_token,
        token_secret => $self->access_secret,
    );

    my $content;
    if ($method =~ m/^(POST|PUT)$/) {
        $hash->{command} ||= $command;
        $content = $self->_template($hash);
        $opts{extra_params} = { xml => $content } if ($method eq 'POST');
    }

    my $request     = Net::OAuth->request("protected resource")->new(%opts);
    my $private_key = Crypt::OpenSSL::RSA->new_private_key($self->cert);
    $request->sign($private_key);

    my $req = HTTP::Request->new($method, $request_url);

    if ($hash and ($method eq 'POST')) {
        $req->content($request->to_post_body);
        $req->header('Content-Type' =>
                'application/x-www-form-urlencoded; charset=utf-8');
    }
    else {
        $req->content($content) if ($hash and ($method eq 'PUT'));
        $req->header(Authorization => $request->to_authorization_header);
    }

    print STDERR $req->as_string if $self->debug;

    my $res = $self->ua->request($req);

    if ($res->is_success) {
        print STDERR "Got Content ", $res->content, "\n" if $self->debug;
        return XMLin($res->content);
    }
    else {
        warn "Something went wrong: " . $res->status_line;
        $self->error($res->status_line . " " . $res->content);
    }

    return;
}

=head2 _template

=cut

sub _template {
    my ($self, $hash) = @_;

    $hash->{command} .= '.tt';
    print STDERR Dumper($hash) if $self->debug;
    my $tt;
    if ($self->debug) {
        $tt = Template->new(
            DEBUG        => 'all',
            INCLUDE_PATH => [ $self->template_path ],
        );
    }
    else {
        $tt = Template->new(INCLUDE_PATH => [ $self->template_path ]);
    }
    my $template = '';
    $tt->process('frame.tt', $hash, \$template)
        || die $tt->error;
    utf8::encode($template);
    print STDERR $template if $self->debug;

    return $template;
}

=head1 AUTHOR

Lenz Gschwendtner, C<< <norbu09 at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-net-xero at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Net-Xero>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Net::Xero


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Net-Xero>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Net-Xero>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Net-Xero>

=item * Search CPAN

L<http://search.cpan.org/dist/Net-Xero/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 COPYRIGHT & LICENSE

Copyright 2010 Lenz Gschwendtner.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


=cut

__PACKAGE__->meta->make_immutable();
